App/src/Service/GroupExpenseBalancer.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace App\Service;
+
+use App\Entity\Bilan;
+use App\Entity\Expense;
+
+class GroupExpenseBalancer
+{
+    public function expenseBalancer($expenses)
+    {
+        /**
+         * @var Bilan[]
+         */
+        $bilans = array_reduce($expenses, static function (array $bilans, Expense $expense) {
+            $payer = $expense->getPayer();
+            if (!array_key_exists($payer, $bilans)) {
+                return [...$bilans, $payer => new Bilan($payer)];
+            }
+
+            return $bilans;
+        }, []);
+
+        $this->setExpenses($expenses, $bilans);
+
+        return $bilans;
+    }
+
+    private function setExpenses($expenses, $bilans)
+    {
+        foreach ($expenses as $expense) {
+            $amount = $expense->getAmount();
+            $participants = $expense->getParticipants();
+            $countParticipants = count($participants);
+            $rest = $amount % $countParticipants;
+            $amountByParticipants = ($amount - $rest) / $countParticipants;
+            $payer = $expense->getPayer();
+
+            $this->updateBilan($bilans, $amount, $participants, $payer, $amountByParticipants);
+        }
+    }
+
+    private function updateBilan($bilans, $amount, $participants, $payer, $amountByParticipants)
+    {
+        foreach ($bilans as $bilan) {
+            $name = $bilan->getName();
+            $cost = $bilan->getCost();
```

Comment:
This variable is used a single time in case of the payer's bilan, maybe you can have a function like in the bilan that add a cost instead of set it.

Reply:
```
foreach ($bilans as $bilan) {
            $name = $bilan->getName();
            $participations = $bilan->getParticipation();
            $owe = $bilan->getOwe();

            if ($payer === $name) {
                $bilan->setCost($bilan->getCost() + $amount);
            }
```

Id write this to avoid set the $cost